File: src/Phan_04/inheritance.py
# Base or Super Class
class Animal:
    def __init__(self, name, age):
        self.name = name
        self.age = age
        print("Animal constructor")

    def __del__(self):
        print("Animal destructor")

    def description(self):
        print(f"Animal {self.name} is {self.age} years old")


# Inherited or Subclass
class Dog(Animal):
    def __init__(self, name, age, breed):
        super().__init__(name, age)

        # new attributes
        self.breed = breed
        # Animal.__init__(self, name, age) # the same to super()
        print("Dog constructor")

    def __del__(self):
        super().__del__()
        # Animal.__del__(self)             # the same to super()
        print("Dog destructor")

    # new methods
    def bark(self):
        print(f"{self.name} is barking")

    # override methods
    def description(self):
        print(f"Dog {self.name} is {self.age} years old and is a {self.breed}")


animal = Animal("Kuro", 5)
animal.description()  # Animal Kuro is 5 years old

dog = Dog("Rocky", 10, "Husky")
dog.description()  # Dog Rocky is 10 years old and is a Husky

File: src/Phan_04/inheritance.cpp
#include<iostream>
using namespace std;

// Base or Super Class
class Animal {
protected:
	string name; 
	int age;
public:
	Animal() {
		name = "";
		age = 0;
		cout << "Animal default constructor\n";
	}
	Animal(string name, int age) {
		this->name = name;
		this->age = age;
		cout << "Animal parameterized constructor\n";
	}
	~Animal() {
		cout << "Animal destructor\n";
	}
	void description() {
		cout << "Animal " << name << " is " << age << " years old\n";
	}
};

// Inherited or Subclass
class Dog : public Animal {
private:
	// new attributes
    string breed;
public: 
	Dog() {
        breed = "";
		cout << "Dog default constructor\n";
	}
    // inherit the parameterized constructor from the parent
    Dog(string name, int age, string breed) : Animal(name, age){
        this->breed = breed;
        cout << "Dog parameterized constructor\n"; 
    }
	~Dog() {
		cout << "Dog destructor\n";
	}
	// new methods
	void bark() {
		cout << name << " is barking\n";
	}
	// override methods
	void description() {
		cout << "Animal " << name << " is " << age << " years old and is a " << breed << endl;
	}
};

int main() {
	Animal animal("Kuro", 5);
	animal.description();	// Animal Kuro is 5 years old

	Dog dog("Rocky", 10, "Husky");
	dog.description();	// Dog Rocky is 10 years old and is a Husky

	return 0;
}
